lib/chef/provider/aws_network_acl.rb

```diff
@@ -0,0 +1,126 @@
+require 'chef/provisioning/aws_driver/aws_provider'
+require 'chef/resource/aws_vpc'
+require 'retryable'
+
+class Chef::Provider::AwsNetworkAcl < Chef::Provisioning::AWSDriver::AWSProvider
+  def action_create
+    network_acl = super
+
+    apply_rules(network_acl)
+  end
+
+  protected
+
+  def create_aws_object
+    converge_by "create new Network ACL #{new_resource.name} in #{region}" do
+      options = {}
+      options[:vpc] = new_resource.vpc if new_resource.vpc
+      options = AWSResource.lookup_options(options, resource: new_resource)
+
+      Chef::Log.debug("VPC: #{options[:vpc]}")
+
+      network_acl = new_resource.driver.ec2.network_acls.create(options)
+      Retryable.retryable(:tries => 15, :sleep => 1, :on => AWS::EC2::Errors::InvalidNetworkAclID::NotFound) do
+        network_acl.tags['Name'] = new_resource.name
+      end
+      network_acl
+    end
+  end
+
+  def update_aws_object(network_acl)
+    if !new_resource.vpc.nil?
+      desired_vpc = Chef::Resource::AwsVpc.get_aws_object_id(new_resource.vpc, resource: new_resource)
+      if desired_vpc != network_acl.vpc_id
+        raise "Network ACL VPC cannot be changed after being created!  Desired VPC for #{new_resource.to_s} was #{new_resource.vpc} (#{desired_vpc}) and actual VPC is #{network_acl.vpc_id}"
+      end
+    end
+  end
+
+  def destroy_aws_object(network_acl)
+    # TODO if purging, do we need to destory the linked subnets?
+    converge_by "delete #{new_resource.to_s} in #{region}" do
+      network_acl.delete
+    end
+  end
+
+  private
+
+  def apply_rules(network_acl)
+    current_rules = network_acl.entries.map { |entry| entry_to_hash(entry) }
+    inbound_rules = new_resource.inbound_rules
+    outbound_rules = new_resource.outbound_rules
+    # AWS requires a deny all rule at the end. Delete here so we don't
+    # try to compare.
+    current_rules.delete_if { |rule| rule[:rule_number] == 32767 }
+
+    current_inbound_rules = current_rules.select { |rule| rule[:egress] == false }
+    desired_inbound_rules = inbound_rules ? inbound_rules.map { |rule| rule[:egress] = false; rule } : []
```

Comment:
One thing we should check for here is the difference between not specifying the `inbound_rules` attribute and specifying it as `inbound_rules []`.  This has come up recently on the Security Group resource.

In the first case (not specified) we don't change existing inbound rules at all.  In the second case (empty) we delete all inbound rules.